### PR TITLE
[video dataset]expose more arguments of VideoClips in video datasets

### PR DESCRIPTION
**Pull request by stephenyan1231**
- In `__init__()` method of class `VideoClips`, two arguments `frame_rate` and `_precomputed_metadata` are not yet exposed in dataset classes for hmdb51, ucf101 and kinetics400.  This PR updates dataset classes to expose those 2 arguments.
- Add a class `VisionVideoDataset` to abstract out API `get_metadata()`. The caller can retrieve video dataset metadata and decide how to cache it. This is dependent on a previous PR (https://github.com/pytorch/vision/pull/1303/)

### CODE REVIEW

File: torchvision/datasets/hmdb51.py [hunk 1]
```diff
@@ -64,7 +65,17 @@ def __init__(self, root, annotation_path, frames_per_clip, step_between_clips=1,
...
-        video_clips = VideoClips(video_list, frames_per_clip, step_between_clips)
+        video_clips = VideoClips(
+            video_list,
+            frames_per_clip,
+            step_between_clips,
+            frame_rate,
+            precomputed_metadata,
+            precomputed_metadata_filepath,
+        )
+        if save_metadata_filepath:
```
Comment (fmassa) [inline torchvision/datasets/hmdb51.py]: I'm not sure if it should be up to the dataset to save the metadata, or the training code around it.
I'd say that this caching should be part of the training pipeline. Thoughts?

File: torchvision/datasets/hmdb51.py [hunk 2]
```diff
@@ -64,7 +65,17 @@ def __init__(self, root, annotation_path, frames_per_clip, step_between_clips=1,
...
         self.classes = classes
         video_list = [x[0] for x in self.samples]
-        video_clips = VideoClips(video_list, frames_per_clip, step_between_clips)
+        video_clips = VideoClips(
+            video_list,
+            frames_per_clip,
+            step_between_clips,
+            frame_rate,
+            precomputed_metadata,
+            precomputed_metadata_filepath,
```
Comment (fmassa) [inline torchvision/datasets/hmdb51.py]: I'm not sure it's a good idea to have the API supporting `precomputed_metadata` and `precomputed_metadata_filepath`, I think that it should support only one of the use-cases, which can be obtained by the other easily. But this is unrelated to this PR, and is dependent on a different PR, so let's move this part of the discussion elsewhere

File: torchvision/datasets/hmdb51.py [hunk 3]
```diff
@@ -50,7 +50,8 @@ class HMDB51(VisionDataset):
     }
 
     def __init__(self, root, annotation_path, frames_per_clip, step_between_clips=1,
-                 fold=1, train=True, transform=None):
+                 frame_rate=None, precomputed_metadata=None, precomputed_metadata_filepath=None,
```
Comment (fmassa) [inline torchvision/datasets/hmdb51.py]: I would be ok exposing the `_precomputed_metadata` as a private argument in the constructor for now, and also expose the `frame_rate`, but I think that the `precomputed_metadata_filepath` and `save_metadata_filepath` should live outside of the dataset and be handled by the training code.

File: torchvision/datasets/hmdb51.py (hunk 1 above)
Comment (stephenyan1231) [reply torchvision/datasets/hmdb51.py]: @fmassa I agree with you. I will remove optional arguments `precomputed_metadata_filepath`, and `save_metadata_filepath`.

File: torchvision/datasets/hmdb51.py (hunk 3 above)
Comment (stephenyan1231) [reply torchvision/datasets/hmdb51.py]: I agree. @fmassa

**Comment by fmassa** [comment]
One thing that bothers me with the current approach is that there is no validation whatsoever that the `precomputed_metadata` and the `samples` (that have been computed from the files) do actually match anyhow.

I see a few possibilities:
- enrich the `metadata` field from `Dataset` to contain `samples` as well (or some function of it), and assert that they have the same number of elements as `video_clips` (and make it start with an underscore, so `_precomputed_metadata` instead of `precomputed_metadata`
- entirely pickle the Dataset instead of adding support for it to construct from metadata. leave to the training code to validate that they actually match
- add a `@classmethod` called `from_metadata` that constructs dataset from the metadata. This keeps the constructor simpler, but will also require that we store more information in the `metadata`, probably the `root`, `classes` and `samples` features. In this case, the `from_metadata` classmethod would take as arguments `metadata`, `transforms` and maybe `frames_per_clip` and `step_between_clips`.

Thoughts?

**Comment by stephenyan1231** [comment]
> One thing that bothers me with the current approach is that there is no validation whatsoever that the `precomputed_metadata` and the `samples` (that have been computed from the files) do actually match anyhow.
> 
> I see a few possibilities:
> 
> * enrich the `metadata` field from `Dataset` to contain `samples` as well (or some function of it), and assert that they have the same number of elements as `video_clips` (and make it start with an underscore, so `_precomputed_metadata` instead of `precomputed_metadata`
> * entirely pickle the Dataset instead of adding support for it to construct from metadata. leave to the training code to validate that they actually match
> * add a `@classmethod` called `from_metadata` that constructs dataset from the metadata. This keeps the constructor simpler, but will also require that we store more information in the `metadata`, probably the `root`, `classes` and `samples` features. In this case, the `from_metadata` classmethod would take as arguments `metadata`, `transforms` and maybe `frames_per_clip` and `step_between_clips`.
> 
> Thoughts?

**Comment by stephenyan1231** [comment]
Since both PR 1303 and current PR 1310 modify common files such as `video_utils.py`, I close current PR 1310, and merge changes into PR 1303 for easier develop and code review. @fmassa
